packages/apps/plugins/plugin-script/src/components/ScriptEditor/ScriptEditor.tsx

```diff
@@ -2,56 +2,137 @@
+  const checkImports =
+    ts &&
+    debounce((state: EditorState) => {
+      syntaxTree(state).iterate({
+        enter: (node) => {
+          if (node.name === 'ImportDeclaration') {
+            const [text] = state.doc.slice(node.from, node.to);
+            void ts.loadImport(text);
+          }
+        },
+      });
+    }, 500);
 
-export const ScriptEditor = ({ source, themeMode, className }: ScriptEditorProps) => {
   const extensions = useMemo(
     () => [
       // TODO(burdon): Use basic set-up (e.g., bracket matching).
-      // TODO(burdon): Use this in text editor (cancels highlight current line)
-      // basicSetup,
       minimalSetup,
-
       lineNumbers(),
+
       // TODO(burdon): Syntax highlighting theme.
-      javascript({ typescript: true }),
-      autocompletion({ activateOnTyping: false }),
-      keymap.of([...completionKeymap]),
+      javascript({
+        typescript: true,
+        jsx: true,
+      }),
+
+      // https://github.com/val-town/codemirror-ts
+      // TODO(burdon): Create worker (configure vite; need extension for plugin architecture?)
+      //  https://github.com/val-town/codemirror-ts?tab=readme-ov-file#setup-worker
+      ts && path
+        ? [
+            tsFacet.of({ env: ts.env, path }),
+            tsSync(),
+            tsLinter(),
+            tsHover({
+              renderTooltip: (info: HoverInfo) => {
+                const div = document.createElement('div');
+                if (info.quickInfo?.displayParts) {
+                  for (const part of info.quickInfo.displayParts) {
+                    // console.log(part.kind);
+                    const span = div.appendChild(document.createElement('span'));
+                    span.className = `cm-tooltip-info cm-tooltip-info-${part.kind}`;
+                    span.innerText = part.text;
+                  }
+                }
+
+                return { dom: div };
+              },
+            }),
+            styles,
+            autocompletion({
+              override: [tsAutocomplete()],
+            }),
+            // Dynamically import types.
+            checkImports
+              ? EditorView.updateListener.of(({ state, changes }) => {
+                  if (!changes.empty) {
+                    checkImports(state);
+                  }
+                })
+              : [],
+          ]
+        : [],
 
       EditorView.baseTheme(defaultTheme),
       EditorView.darkTheme.of(themeMode === 'dark'),
       EditorView.contentAttributes.of({ class: '!px-2' }),
 
-      // TODO(burdon): Presence.
+      // TODO(burdon): Add presence.
       automerge(source),
     ],
-    [source, themeMode],
+    [source, themeMode, ts, path],
   );
 
-  const { parentRef } = useTextEditor(
+  const { parentRef, view } = useTextEditor(
     {
       extensions,
       doc: DocAccessor.getValue(source),
     },
     [extensions],
   );
 
+  useEffect(() => {
```

Comment:
wouldn't you wanna run this on every edit (with thottling) instead of only on load?

Reply:
the custom extension listener above does this. this effect is just for on start.